File: sdk/nodejs1/sdk/src/zrok/share.ts
import {Root} from "./environment";
import {
    ShareApi,
    ShareRequest as ApiShareRequest,
    ShareRequestBackendModeEnum,
    ShareRequestShareModeEnum,
    UnshareRequest
} from "../api";

export type ShareMode = string;
export const PRIVATE_SHARE_MODE: ShareMode = "private";
export const PUBLIC_SHARE_MODE: ShareMode = "public";

export type BackendMode = string;
export const PROXY_BACKEND_MODE: BackendMode = "proxy";
export const TCP_TUNNEL_BACKEND_MODE: BackendMode = "tcpTunnel";
export const UDP_TUNNEL_BACKEND_MODE: BackendMode = "udpTunnel";

export type PermissionMode = string;
export const OPEN_PERMISSION_MODE = "open";
export const CLOSED_PERMISSION_MODE = "closed";

export class ShareRequest {
    reserved: boolean;
    uniqueName: string|undefined;
    backendMode: BackendMode;
    shareMode: ShareMode;
    target: string;
    frontends: string[]|undefined;
    basicAuth: string[]|undefined;
    oauthProvider: string|undefined;
    oauthEmailAddressPatterns: string[]|undefined;
    oauthAuthorizationCheckInterval: string|undefined;
    permissionMode: PermissionMode;
    accessGrants: string[]|undefined;

    constructor(shareMode: ShareMode, backendMode: BackendMode, target: string) {
        this.reserved = false;
        this.uniqueName = undefined;
        this.backendMode = backendMode;
        this.shareMode = shareMode;
        this.target = target;
        this.frontends = undefined;
        this.basicAuth = undefined;
        this.oauthProvider = undefined;
        this.oauthEmailAddressPatterns = undefined;
        this.oauthAuthorizationCheckInterval = undefined;
        this.permissionMode = CLOSED_PERMISSION_MODE;
        this.accessGrants = undefined;
    }
}

export class Share {
    shareToken: string;
    frontendEndpoints: string[]|undefined;

    constructor(shareToken: string, frontendEndpoints: string[]|undefined) {
        this.shareToken = shareToken;
        this.frontendEndpoints = frontendEndpoints;
    }
}

export const createShare = async (root: Root, request: ShareRequest): Promise<Share> => {
    if(!root.isEnabled()) {
        throw new Error("environment is not enabled; enable with 'zrok enable' first!");
    }

    let req: ApiShareRequest;
    switch(request.shareMode) {
        case PRIVATE_SHARE_MODE:
            req = toPrivateApiShareRequest(root, request);
            break;
        case PUBLIC_SHARE_MODE:
            req = toPublicApiShareRequest(root, request);
            break;
        default:
            throw new Error("unknown share mode '" + request.shareMode + "'");
    }

    let shr = await new ShareApi(root.apiConfiguration()).share({body: req})
        .catch(resp => {
            throw new Error("unable to create share: " + resp);
        });

    return new Share(shr.shareToken!, shr.frontendProxyEndpoints);
}

export const deleteShare = (root: Root, shr: Share): void => {
    if(!root.isEnabled()) {
        throw new Error("environment is not enable; enable with 'zrok enable' first!");
    }
    let req: UnshareRequest = {
        envZId: root.environment?.zId!,
        shareToken: shr.shareToken
    };
    new ShareApi(root.apiConfiguration()).unshare({body: req})
        .catch(resp => {
            throw new Error("unable to delete share: " + resp);
        });
}

const toPrivateApiShareRequest = (root: Root, request: ShareRequest): ApiShareRequest => {
    return {
        envZId: root.environment?.zId,
        shareMode: ShareRequestShareModeEnum.Private,
        backendMode: toApiBackendMode(request.backendMode),
        backendProxyEndpoint: request.target,
    };
}

const toPublicApiShareRequest = (root: Root, request: ShareRequest): ApiShareRequest => {
    return {
        envZId: root.environment?.zId,
        shareMode: ShareRequestShareModeEnum.Public,
        frontendSelection: request.frontends,
        backendMode: toApiBackendMode(request.backendMode),
        backendProxyEndpoint: request.target,
        authScheme: "none",
    };
}

const toApiBackendMode = (mode: BackendMode): ShareRequestBackendModeEnum|undefined => {
    switch(mode) {
        case PROXY_BACKEND_MODE:
            return ShareRequestBackendModeEnum.Proxy;
        case TCP_TUNNEL_BACKEND_MODE:
            return ShareRequestBackendModeEnum.TcpTunnel;
        case UDP_TUNNEL_BACKEND_MODE:
            return ShareRequestBackendModeEnum.UdpTunnel;
        default:
            return undefined;
    }
}

File: sdk/golang/sdk/share.go
package sdk

import (
	"strings"

	httptransport "github.com/go-openapi/runtime/client"
	"github.com/openziti/zrok/environment/env_core"
	"github.com/openziti/zrok/rest_client_zrok/share"
	"github.com/openziti/zrok/rest_model_zrok"
	"github.com/pkg/errors"
)

func CreateShare(root env_core.Root, request *ShareRequest) (*Share, error) {
	if !root.IsEnabled() {
		return nil, errors.New("environment is not enabled; enable with 'zrok enable' first!")
	}

	var err error
	var out *share.ShareParams

	switch request.ShareMode {
	case PrivateShareMode:
		out = newPrivateShare(root, request)
	case PublicShareMode:
		out = newPublicShare(root, request)
	default:
		return nil, errors.Errorf("unknown share mode '%v'", request.ShareMode)
	}

	if len(request.BasicAuth) > 0 {
		out.Body.AuthScheme = string(Basic)
		for _, basicAuthUser := range request.BasicAuth {
			tokens := strings.Split(basicAuthUser, ":")
			if len(tokens) == 2 {
				out.Body.BasicAuthUsers = append(out.Body.BasicAuthUsers, &rest_model_zrok.AuthUser{Username: strings.TrimSpace(tokens[0]), Password: strings.TrimSpace(tokens[1])})
			} else {
				return nil, errors.Errorf("invalid username:password '%v'", basicAuthUser)
			}
		}
	}

	if request.OauthProvider != "" {
		out.Body.AuthScheme = string(Oauth)
	}

	zrok, err := root.Client()
	if err != nil {
		return nil, errors.Wrap(err, "error getting zrok client")
	}
	auth := httptransport.APIKeyAuth("X-TOKEN", "header", root.Environment().AccountToken)

	in, err := zrok.Share.Share(out, auth)
	if err != nil {
		return nil, errors.Wrap(err, "unable to create share")
	}

	return &Share{
		Token:             in.Payload.ShareToken,
		FrontendEndpoints: in.Payload.FrontendProxyEndpoints,
	}, nil
}

func newPrivateShare(root env_core.Root, request *ShareRequest) *share.ShareParams {
	req := share.NewShareParams()
	req.Body = &rest_model_zrok.ShareRequest{
		EnvZID:            root.Environment().ZitiIdentity,
		ShareMode:         string(request.ShareMode),
		PrivateShareToken: request.PrivateShareToken,
		BackendMode:       string(request.BackendMode),
		Target:            request.Target,
		AuthScheme:        string(None),
		PermissionMode:    string(request.PermissionMode),
		AccessGrants:      request.AccessGrants,
	}
	return req
}

func newPublicShare(root env_core.Root, request *ShareRequest) *share.ShareParams {
	req := share.NewShareParams()
	req.Body = &rest_model_zrok.ShareRequest{
		EnvZID:               root.Environment().ZitiIdentity,
		ShareMode:            string(request.ShareMode),
		BackendMode:          string(request.BackendMode),
		Target:               request.Target,
		AuthScheme:           string(None),
		OauthEmailDomains:    request.OauthEmailAddressPatterns,
		OauthProvider:        request.OauthProvider,
		OauthRefreshInterval: request.OauthRefreshInterval.String(),
		PermissionMode:       string(request.PermissionMode),
		AccessGrants:         request.AccessGrants,
	}
	for _, nss := range request.NamespaceSelections {
		req.Body.NamespaceSelections = append(req.Body.NamespaceSelections, &rest_model_zrok.NamespaceSelection{
			NamespaceToken: nss.NamespaceToken,
			Name:           nss.Name,
		})
	}
	return req
}

func DeleteShare(root env_core.Root, shr *Share) error {
	req := share.NewUnshareParams()
	req.Body.EnvZID = root.Environment().ZitiIdentity
	req.Body.ShareToken = shr.Token

	zrok, err := root.Client()
	if err != nil {
		return errors.Wrap(err, "error getting zrok client")
	}
	auth := httptransport.APIKeyAuth("X-TOKEN", "header", root.Environment().AccountToken)

	_, err = zrok.Share.Unshare(req, auth)
	if err != nil {
		return errors.Wrap(err, "error deleting share")
	}

	return nil
}
